File: new-resource-detector/src/test/java/nl/haploid/octowight/service/ResourceDetectorsServiceIT.java
package nl.haploid.octowight.service;

import mockit.Deencapsulation;
import mockit.StrictExpectations;
import nl.haploid.octowight.AbstractIT;
import nl.haploid.octowight.AtomChangeEvent;
import nl.haploid.octowight.AtomGroup;
import nl.haploid.octowight.TestData;
import nl.haploid.octowight.detector.MockResourceDetector;
import nl.haploid.octowight.detector.ResourceDetector;
import nl.haploid.octowight.registry.data.ResourceRoot;
import org.junit.Before;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.annotation.DirtiesContext;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;

@DirtiesContext
public class ResourceDetectorsServiceIT extends AbstractIT {

	@Autowired
	private ResourceDetectorsService service;

	@Autowired
	private MockResourceDetector mockDetector;

	@Before
	public void setUp() throws Exception {
		Deencapsulation.setField(service, "detectors", Collections.singletonList(mockDetector));
	}

	@Test
	public void testGetDetectorsForAtomType() {
		final List<MockResourceDetector> expectedDetectors = Collections.singletonList(mockDetector);
		final String atomType = "kinsey";
		new StrictExpectations(mockDetector) {{
			mockDetector.getAtomTypes();
			times = 1;
			result = Arrays.asList("crane", atomType);
		}};
		final List<ResourceDetector> actualDetectors = service.getDetectorsForAtomType(atomType);
		assertEquals(expectedDetectors, actualDetectors);
	}

	@Test
	public void testDetectResources() {
		final String atomType = "harris";
		final AtomChangeEvent event1 = TestData.atomChangeEvent(atomType);
		final AtomChangeEvent event2 = TestData.atomChangeEvent("calvet");
		final AtomGroup atomGroup = new AtomGroup();
		atomGroup.setAtomOrigin(event1.getAtomOrigin());
		atomGroup.setAtomType(event1.getAtomType());
		final List<AtomChangeEvent> events = Arrays.asList(event1, event2);
		final List<ResourceRoot> expectedResourceRoots = Collections.singletonList(TestData.resourceRoot(96l));
		new StrictExpectations(mockDetector) {{
			mockDetector.getAtomTypes();
			times = 1;
			result = Arrays.asList("holloway", atomType);
			mockDetector.detect(events);
			times = 1;
			result = expectedResourceRoots;
		}};
		final List<ResourceRoot> actualResourceRoots = service.detectResources(atomGroup, events);
		assertEquals(expectedResourceRoots, actualResourceRoots);
	}
}

File: new-resource-detector/src/test/scala/nl/haploid/octowight/service/ResourceDetectorsServiceIT.scala
package nl.haploid.octowight.service

import java.util

import nl.haploid.octowight._
import nl.haploid.octowight.detector.ResourceDetector
import org.scalatest.mock.EasyMockSugar
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.util.ReflectionTestUtils

@DirtiesContext
class ResourceDetectorsServiceIT extends AbstractIT with EasyMockSugar with EasyMockInjection {
  @Autowired private[this] val resourceDetectorsService: ResourceDetectorsService = null
  @Mocked private[this] val mockDetector: ResourceDetector = null

  behavior of "Resource detectors service"

  override def beforeEach() = {
    super.beforeEach()
    injectMocks()
    ReflectionTestUtils.setField(resourceDetectorsService, "_detectors", util.Collections.singletonList(mockDetector))
  }

  it should "get detectors for atom category" in {
    val expectedDetectors = List(mockDetector)
    val atomCategory = "kinsey"
    expecting {
      mockDetector.atomCategories andReturn List("crane", atomCategory) once()
    }
    whenExecuting(mockDetector) {
      val actualDetectors = resourceDetectorsService.detectorsForAtomCategory(atomCategory)
      actualDetectors should be(expectedDetectors)
    }
  }

  it should "detect resources" in {
    val atomCategory = "harris"
    val event1 = newresourcedetector.TestData.atomChangeEvent(atomCategory)
    val event2 = newresourcedetector.TestData.atomChangeEvent("calvet")
    val atomGroup = new AtomGroup(origin = event1.atomOrigin, category = event1.atomCategory)
    val events = List(event1, event2)
    val expectedResourceRoots = List(newresourcedetector.TestData.resourceRoot(newresourcedetector.TestData.nextLong))
    expecting {
      mockDetector.atomCategories andReturn List("holloway", atomCategory) once()
      mockDetector.detect(events) andReturn expectedResourceRoots once()
    }
    whenExecuting(mockDetector) {
      val actualResourceRoots = resourceDetectorsService.detectResources(atomGroup, events)
      actualResourceRoots should be(expectedResourceRoots)
    }
  }
}
